### PR TITLE
Require calling a _BoundMethodProxy to get the underlying callable.

### PR DESCRIPTION
**Pull request by anntzer**
Code that used to call `_proxy(...)` now needs to call `_proxy()(...)`.

Instead of catching a ReferenceError, that could be either raised by a
failure to dereference the proxy, or by the underlying callable, one can
now check that `_proxy()` does not return None (to distinguish between
the two cases).

This is the same design as the stdlib's WeakMethod.

See https://github.com/matplotlib/matplotlib/pull/9063#discussion_r134396186.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is PEP 8 compliant 
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by anntzer** [comment]
rewritten to use weakref.WeakMethod directly

**Comment by anntzer** [comment]
rebased

File: lib/matplotlib/cbook/__init__.py [hunk 1]
```diff
@@ -236,18 +162,16 @@ def __setstate__(self, state):
     def connect(self, s, func):
         """Register *func* to be called when signal *s* is generated.
         """
-        self._func_cid_map.setdefault(s, WeakKeyDictionary())
-        # Note proxy not needed in python 3.
-        # TODO rewrite this when support for python2.x gets dropped.
-        proxy = _BoundMethodProxy(func)
+        self._func_cid_map.setdefault(s, {})
```
Comment (tacaswell) [inline lib/matplotlib/cbook/__init__.py]: pros / cons of using `ismthod` vs `try:....except:...` on attempting to create the `WeakMethod`?

Comment (anntzer) [reply lib/matplotlib/cbook/__init__.py]: Edited accordingly.
As mentioned elsewhere I believe quite strongly that the use of weakmethod semantics is pretty silly anyways...

**Comment by tacaswell** [comment]
Circle builds the commit on the branch (rather than the commit from doing the merge) so those won't fix them selves.

File: doc/api/next_api_changes/2018-05-06-AL-callbackregistry.rst [hunk 2]
```diff
@@ -0,0 +1,6 @@
+`CallbackRegistry` now stores callbacks using stdlib's `WeakMethods`
+````````````````````````````````````````````````````````````````````
+
+In particular, this implies that ``CallbackRegistry.callbacks[signal]`` is now
+a mapping of callback ids to `WeakMethods` (i.e., they need to be first called
```
[the remark this answers is not in the document]
Comment (anntzer) [reply doc/api/next_api_changes/2018-05-06-AL-callbackregistry.rst]: fixed x2

File: lib/matplotlib/cbook/__init__.py [hunk 3]
```diff
@@ -61,100 +62,26 @@ def unicode_safe(s):
...
-                return self.func == other.func and self.inst() == other.inst()
-        except Exception:
-            return False
+        return isinstance(other, _StrongRef) and self._obj == other._obj
 
     def __hash__(self):
-        return self._hash
-
-
-def _exception_printer(exc):
```
Comment (tacaswell) [inline lib/matplotlib/cbook/__init__.py]: Why don't we need this anymore?

Comment (anntzer) [reply lib/matplotlib/cbook/__init__.py]: It's still there (line 65) and unchanged, the diff is just confused.

Comment (tacaswell) [reply lib/matplotlib/cbook/__init__.py]: Ah, great sorry for the noise.

**Pull request merged**